src/components/RoiCalculator.vue

```diff
@@ -0,0 +1,432 @@
+import { KlayIconSwitch } from '~klay-icons'
+import { SModal } from '@soramitsu-ui/ui'
+import BigNumber from 'bignumber.js'
+import { RoiType, Tab } from '@/types'
+
+const Period = {
+  d1: '1D',
+  d7: '7D',
+  d14: '14D',
+  d30: '30D',
+  y1: '1Y',
+  y5: '5Y'
+} as const
+
+type Period = typeof Period[keyof typeof Period]
+
+const StakeTabs = {
+  d1: Period.d1,
+  d7: Period.d7,
+  d30: Period.d30,
+  y1: Period.y1,
+  y5: Period.y5
+} as const
+
+type StakeTabs = typeof StakeTabs[keyof typeof StakeTabs]
+
+const CompoundingTabs = {
+  d1: Period.d1,
+  d7: Period.d7,
+  d14: Period.d14,
+  d30: Period.d30
+} as const
+
+type CompoundingTabs = typeof CompoundingTabs[keyof typeof CompoundingTabs]
+
+const StakeUnits = {
+  tokens: 'tokens',
+  USD: 'USD',
+} as const
+
+type StakeUnits = typeof StakeUnits[keyof typeof StakeUnits]
+
+const { t } = useI18n()
+const vBem = useBemClass()
+
+const props = defineProps<{
+  show: boolean
+  type: RoiType
+  apr: BigNumber
+  lpApr?: BigNumber
+  staked: BigNumber
+  balance: BigNumber
+  stakeTokenPrice: BigNumber
+  stakeTokenDecimals: number
+  rewardTokenDecimals: number
+  stakeTokenSymbol: string
+  rewardTokenSymbol: string
+}>()
+const {
+  type,
+  apr,
+  lpApr,
+  staked,
+  balance,
+  stakeTokenPrice,
+  stakeTokenDecimals,
+  rewardTokenDecimals,
+  stakeTokenSymbol,
+  rewardTokenSymbol
+} = toRefs(props)
+const emit = defineEmits<(e: 'update:show', value: boolean) => void>()
+
+const showModel = useVModel(props, 'show', emit)
+
+function makeTabsArray(data: string[]): Tab[] {
+  return data.map(item => ({
+    id: item,
+    label: item
+  }))
+}
+
+const stakeTabs = readonly(makeTabsArray(Object.values(StakeTabs)))
+const compoundingTabs = readonly(makeTabsArray(Object.values(CompoundingTabs)))
+
+const stakeValueInputWrapper = ref<HTMLElement | null>(null)
+const stakeFor = ref(StakeTabs.y1)
+const compoundingEnabled = ref(true)
+const compoundingEvery = ref(CompoundingTabs.d1)
+const stakeValueRaw = ref('$0')
+const parsedStakeValue = ref(new BigNumber(0))
+const stakeUnits = ref<StakeUnits>(StakeUnits.USD)
+
+function setStakeValueInUSD(amount: number) {
+  if (stakeUnits.value === StakeUnits.USD)
+    updateStakeValueRaw(String(amount))
+  else
+    updateStakeValueRaw(new BigNumber(new BigNumber(amount).div(stakeTokenPrice.value).toFixed(stakeTokenDecimals.value)).toString())
+}
+
+function setStakeValueWithBalance() {
+  if (stakeUnits.value === StakeUnits.USD)
+    updateStakeValueRaw(new BigNumber(balance.value.times(stakeTokenPrice.value).toFixed(2)).toString())
+  else
+    updateStakeValueRaw( balance.value.toString())
+}
+
+const totalApr = computed(() => {
+  return apr.value.plus(lpApr?.value ? lpApr.value : 0)
+})
+
+function getPeriodDays(period: Period): number {
+  return {
+    [Period.d1]: 1,
+    [Period.d7]: 7,
+    [Period.d14]: 14,
+    [Period.d30]: 30,
+    [Period.y1]: 365,
+    [Period.y5]: 365 * 5
+  }[period]
+}
```

Comment:
Inefficient each-invokation-object-creation. Also it can be moved out to some const module.

Reply:
Done